New Pokemon Class changes

ZachSellman opened the pull request:
Added Pokemon_class.py, consolidated main and scanner since new "main" purpose is to generate pokemon objects. top_five module added to pull just the top five most mentioned results.

ZachSellman committed:
New Pokemon Class changes

Added Pokemon_class.py, consolidated main and scanner since new "main" purpose is to generate pokemon objects. top_five module added to pull just the top five most mentioned results.

ZachSellman committed:
Update pylint.yml

justincely commented on top_five.py:
```diff
@@ -0,0 +1,17 @@
+def get_top_five(n: dict):
```
why is this one split out here?

ZachSellman replied:
I'm still trying to figure out best practice for this sort of thing. I was thinking since main.py essentially gets the data for formats it into a dictionary of names and objects that maybe whatever is then done TO that data should be separate? Like if I decided to do a top ten, top fifteen, etc. shouldn't those options be in a separate module?

I guess I'm foggy on how MUCH code should be going into each .py file?

justincely replied:
honestly, just keep them all together until you have a reason to split them.  You'll see the natural split when you've got more than this.  Right now this is just small enough for it all to be in one

justincely commented on pokemon_scanner.py:
```diff
@@ -1,69 +0,0 @@
-import praw
```
nice, i like this deletion.  There wasn't really much use in the two files; not much needed separation.

justincely commented on main.py:
```diff
@@ -1,19 +1,88 @@
+
+dotenv.load_dotenv()
+
+client_secret = getenv("CLIENT_SECRET")
```
you'll start seeing this in your linters - but it's good practice in the PEP8 style guide to have all global variables in all upper-case.  so `CLIENT_SECRET` and `USERNAME` and ... etc

justincely commented on Pokemon_class.py:
```diff
@@ -0,0 +1,36 @@
+    def __init__(self, name: str):
+        """constructor method"""
+        self.name = name
+        self.mentions = 1
```
init this to 0, and then increment the mention when it's actually mentioned.  Think about this in the general case - there may be tons of times you'll instantiate a pokemon object for other purposes (testing, setting attributes, maybe getting things ready for later), so don't assume it's created just for the purposes of a mention.  

Also, see the code later for why - doing it starting with 1 is actually harder.

justincely commented on main.py:
```diff
@@ -1,19 +1,88 @@
+    :type sub_id: str
+    :returns: None
+    """
+    if name in mentioned:
+        mentioned[name].add_mention()
+        if comment_id:
+            mentioned[name].add_comment_id(comment_id)
+        else:
+            mentioned[name].add_submission_id(sub_id)
+
+    else:
+        mentioned[name] = Pokemon(name)
+        if comment_id:
+            mentioned[name].add_comment_id(comment_id)
+        else:
+            mentioned[name].add_submission_id(sub_id)
```
This works, but there's a more elegant way to do this that is less prone to breaking.    Basically you check if it hasn't been mentioned/initialized in the past; if not - correct that init, then proceed with just one branch of code now that you know it will always be there.  

```python
if name not in mentioned:
  mentioned[name] = Pokemon(name)

mentioned[name].add_mention()
if comment_id:
   mentioned[name].add_comment_id(comment_id)
else:
   mentioned[name].add_submission_id(sub_id)
 ```
```

ZachSellman replied:
Awesome, that's way better.

justincely commented on main.py:
```diff
@@ -1,19 +1,88 @@
+
+    else:
+        mentioned[name] = Pokemon(name)
+        if comment_id:
+            mentioned[name].add_comment_id(comment_id)
+        else:
+            mentioned[name].add_submission_id(sub_id)
```
are these actually mutually exclusive?  it will either have a comment ID xor a submission id?

ZachSellman replied:
PRAW documentation states there is a comment-specific ID and a submission ID. That said, I haven't actually attempted to get the comment or submission using the ID's yet.

ZachSellman replied:
Oh, maybe to clarify too; I'm counting each "mention" as essentially user(x) mentioned it in either a post title/selftext or in a comment. So one submission could have as many mentions as there are comments + the submission itself if everyone mentions a pokemon in the list.

justincely replied:
gotcha

ZachSellman committed:
Pylint changes, code cleanup

Lots of formatting fixes and discussed code cleanup.

ZachSellman committed:
More Pylint changes

ZachSellman committed:
Update top_five.py

snake_case variable rename

justincely commented:
Also, don't open these in draft if you want review.  Draft means pre-review, so if you want someone to check it out but not do a detailed review because you're still working on it.  Or perhaps you're opening it but it's blocked by something else so you for sure don't want it merged yet.

ZachSellman commented:
I've fixed almost all of the issues Pylint is finding, but it keeps saying it can't import PRAW and some other libraries. I'm not sure how to resolve that?

ZachSellman committed:
consolidated top_five and main

justincely commented:
github actions is actually taking your repo, cloning it on a remote computer, and starting to run the commands you specified.    You haven't told it to install your dependencies yet.  you need a line like `pip install -r requirements.txt` in there so it actually installs PRAW on the local system

ZachSellman committed:
Update pylint.yml

added line to install requirements

ZachSellman committed:
Rename Pokemon_class.py to pokemon_class.py

ZachSellman committed:
Update main.py

Added module doc string. Unsure if this is actually necessary since main.py would typically be run from the command line?

justincely reviewed (approved):
nice, checks pass, happy faces all around 🚀

ZachSellman commented:
It won't let me merge it currently... says branch has conflicts at the pylint.yml file that must be resolved, but resolve button is grey'd out. Tried git status in GitBash but it says the branch is clean... Any ideas?

justincely commented:
haha ugh, yeah we got into gross history state.  Don't worry much about it - when you get the chance to merge, use the down arrow and select ` Squash and merge`.  You'll get a chance to edit the commit message to something better than the crazy history

justincely commented:
@ZachSellman check it over and see if it looks ok.  I resolved the conflict manually
